scene/resources/material.cpp

```diff
@@ -700,252 +700,298 @@ void BaseMaterial3D::_update_shader() {
+			code += ", diffuse_lambert";
 			break;
 		case DIFFUSE_LAMBERT_WRAP:
-			code += ",diffuse_lambert_wrap";
+			code += ", diffuse_lambert_wrap";
 			break;
 		case DIFFUSE_TOON:
-			code += ",diffuse_toon";
+			code += ", diffuse_toon";
 			break;
 		case DIFFUSE_MAX:
 			break; // Internal value, skip.
 	}
 	switch (specular_mode) {
 		case SPECULAR_SCHLICK_GGX:
-			code += ",specular_schlick_ggx";
+			code += ", specular_schlick_ggx";
 			break;
 		case SPECULAR_TOON:
-			code += ",specular_toon";
+			code += ", specular_toon";
 			break;
 		case SPECULAR_DISABLED:
-			code += ",specular_disabled";
+			code += ", specular_disabled";
 			break;
 		case SPECULAR_MAX:
 			break; // Internal value, skip.
 	}
 	if (features[FEATURE_SUBSURFACE_SCATTERING] && flags[FLAG_SUBSURFACE_MODE_SKIN]) {
-		code += ",sss_mode_skin";
+		code += ", sss_mode_skin";
 	}
 
 	if (shading_mode == SHADING_MODE_UNSHADED) {
-		code += ",unshaded";
+		code += ", unshaded";
 	}
 	if (flags[FLAG_DISABLE_DEPTH_TEST]) {
-		code += ",depth_test_disabled";
+		code += ", depth_test_disabled";
 	}
 	if (flags[FLAG_PARTICLE_TRAILS_MODE]) {
-		code += ",particle_trails";
+		code += ", particle_trails";
 	}
 	if (shading_mode == SHADING_MODE_PER_VERTEX) {
-		code += ",vertex_lighting";
+		code += ", vertex_lighting";
 	}
 	if (flags[FLAG_DONT_RECEIVE_SHADOWS]) {
-		code += ",shadows_disabled";
+		code += ", shadows_disabled";
 	}
 	if (flags[FLAG_DISABLE_AMBIENT_LIGHT]) {
-		code += ",ambient_light_disabled";
+		code += ", ambient_light_disabled";
 	}
 	if (flags[FLAG_USE_SHADOW_TO_OPACITY]) {
-		code += ",shadow_to_opacity";
+		code += ", shadow_to_opacity";
 	}
 	if (flags[FLAG_DISABLE_FOG]) {
-		code += ",fog_disabled";
+		code += ", fog_disabled";
 	}
 
 	if (transparency == TRANSPARENCY_ALPHA_DEPTH_PRE_PASS) {
-		code += ",depth_prepass_alpha";
+		code += ", depth_prepass_alpha";
 	}
 
-	// Although its technically possible to do alpha antialiasing without using alpha hash or alpha scissor,
+	// Although it's technically possible to do alpha antialiasing without using alpha hash or alpha scissor,
 	// it is restricted in the base material because it has no use, and abusing it with regular Alpha blending can
-	// saturate the MSAA mask
+	// saturate the MSAA mask.
 	if (transparency == TRANSPARENCY_ALPHA_HASH || transparency == TRANSPARENCY_ALPHA_SCISSOR) {
-		// alpha antialiasing is only useful in ALPHA_HASH or ALPHA_SCISSOR
+		// Alpha antialiasing is only useful with ALPHA_HASH or ALPHA_SCISSOR.
 		if (alpha_antialiasing_mode == ALPHA_ANTIALIASING_ALPHA_TO_COVERAGE) {
-			code += ",alpha_to_coverage";
+			code += ", alpha_to_coverage";
 		} else if (alpha_antialiasing_mode == ALPHA_ANTIALIASING_ALPHA_TO_COVERAGE_AND_TO_ONE) {
-			code += ",alpha_to_coverage_and_one";
+			code += ", alpha_to_coverage_and_one";
 		}
 	}
 
-	code += ";\n";
+	code += R"(;
+
+uniform vec4 albedo : source_color;
+uniform sampler2D texture_albedo : source_color, )" +
+			texfilter_str + ";\n";
```

Comment:
I think this is a case where readability suffers a bit.

I'd keep `code += ";\n";` separate as the end of the huge `render_mode` conditional formatting.
Could even put everything from `code += "render_mode ";` to this closing semicolon in a dedicated scope to make it more readable.

Then for the second part, a raw string seems fine, but can we use `vformat()` for those instead of awkward concatenation?